### PR TITLE
Added support for node and npm based projects

### LINKED ISSUE #1017: Add AddNpmApp and AddNodeApp to Aspire.Hosting
We should move this boilerplate to the core (we'll figure out refactorig of these packages later).

https://github.com/dotnet/aspire-samples/blob/main/samples/AspireWithNode/AspireWithNode.AppHost/NodeAppHostingExtensions.cs

cc @DamianEdwards

### PR DESCRIPTION
**Pull request by davidfowl**
- Added AddNodeApp and AddNpmApp extension methods
- Flow the project's working directory into the host in 2 ways, via an attribute and as another project in the list. By default, executable's working directory are relative to the AppHost's project directory (this is what the attribute is used for).
- Added support for flowing the dynamically chosen port to the application by specifying a "portEnvVar" setting on the ServiceBindingAnnotation.
- Added `IResourceWithServiceDiscovery` which allows custom resources to export service discovery information.
- This test relies on node being installed.
- ~Added node.v0 resource for node apps~
- Improve support for executable resources in the manifest.

TODO: 
1. ~Test for npm based project.~
2. Skip tests if node isn't on the path.
3. ~Figure out what to do for deployment? Should we model this as a `node.v0` resource? That would match what we do for .NET projects...~

Fixes https://github.com/dotnet/aspire/issues/1017

PS: Fixing some of these issues (flowing port to the app and fixing working directories makes it easier to execute other project types in general). AddNpmApp and AddNodeApp are very thin wrappers now.

### CODE REVIEW

File: src/Aspire.Hosting/Node/NodeExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    /// <summary>
+    /// Injects service discovery information as environment variables from the project resource into the destination resource, using the source resource's name as the service name.
+    /// Each service binding defined on the project resource will be injected using the format "services__{sourceResourceName}__{bindingIndex}={bindingNameQualifiedUriString}."
+    /// </summary>
+    /// <typeparam name="TDestination">The destination resource.</typeparam>
+    /// <typeparam name="TSource">The source project resource.</typeparam>
+    /// <param name="builder">The resource where the service discovery information will be injected.</param>
+    /// <param name="source">The resource from which to extract service bindings.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{TDestination}"/>.</returns>
+    public static IResourceBuilder<TDestination> WithReference<TDestination, TSource>(this IResourceBuilder<TDestination> builder, IResourceBuilder<TSource> source)
```
Comment (davidfowl) [inline src/Aspire.Hosting/Node/NodeExtensions.cs:59]: @mitchdenny Now might be the time to fix this. Right now, this requires adding explicit overloads for things that support service discovery. We need to make it more contract based.

File: src/Aspire.Hosting/Node/NodeExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        var resource = new NodeAppResource(name, "npm", workingDirectory, allArgs);
+
+        return builder.AddResource(resource)
+                      .WithNodeDefaults();
+    }
+
+    private static IResourceBuilder<NodeAppResource> WithNodeDefaults(this IResourceBuilder<NodeAppResource> builder) =>
+        builder.WithOtlpExporter()
+            .WithEnvironment("NODE_ENV", builder.ApplicationBuilder.Environment.IsDevelopment() ? "development" : "production")
+            .ExcludeFromManifest();
```
Comment (davidfowl) [inline src/Aspire.Hosting/Node/NodeExtensions.cs:53]: This feels bad to bake into the method.

File: src/Aspire.Hosting/ApplicationModel/ServiceBindingAnnotation.cs [hunk 3]
```diff
@@ -15,7 +15,7 @@ namespace Aspire.Hosting.ApplicationModel;
 [DebuggerDisplay("Type = {GetType().Name,nq}, Name = {Name}")]
 public sealed class ServiceBindingAnnotation : IResourceAnnotation
 {
-    public ServiceBindingAnnotation(ProtocolType protocol, string? uriScheme = null, string? transport = null, string? name = null, int? port = null, int? containerPort = null, bool? isExternal = null)
+    public ServiceBindingAnnotation(ProtocolType protocol, string? uriScheme = null, string? transport = null, string? name = null, int? port = null, int? containerPort = null, bool? isExternal = null, string? portEnvVar = null)
```
Comment (mitchdenny) [inline src/Aspire.Hosting/ApplicationModel/ServiceBindingAnnotation.cs:18]: Could ```portEnvVar``` just be ```env``` or ```environmentVariable```?

**Comment by mitchdenny** [comment]
Overall this looks pretty solid.

**Comment by davidfowl** [comment]
> I think if we go down the path of emitting a node.v0 resource that we'll have to give some kind of guidance on what we expect tools to do with it. For example, if azd or aspirate encounter a node.v0 resource should then look for a Dockerfile in the same path as package.json? Would there be an explciit override in package.json to specify the location of the Dockerfile?

Right, I think this is what we need to decide. Today for project.v0 resources (maybe this should be a dotnet.v0 resource) we recommend people use container builds, but that's all dependent on the target environment. You could imagine it meaning do a dotnet publish and zip the results and upload it to the target (It could mean use the nodejs build pack... etc)

I think docker file support needs to be an explicit call.

**Comment by mitchdenny** [comment]
Regarding renaming `project.v0` to `dotnet.v0` ... I **think** this might make sense. Although we'd have to consider how things like `*.esproj` files might be handled if we want to explicitly support them. Maybe AddProject interrogates the project metadata and calls AddDotnet or AddNpm as appropriate.

What I mean about Docker support though is that I think as a framework orientated towards cloud native development we assume that if you have a reosurce like Node or Python apps that there will be a Docker build involved in deployment.

**Review by mitchdenny — APPROVED** [review]
Assuming ```WriteNodeApp``` is temporarily in ```ManifestPublisher``` until the ```ManifestPublisherCallbackAnnotation``` is updated to take a context.

File: src/Aspire.Hosting/build/Aspire.Hosting.targets [hunk 4]
```diff
@@ -63,13 +63,51 @@ public class ]]>%(ClassName)<![CDATA[ : IServiceMetadata
     </ItemGroup>
   </Target>
 
+  <Target Name="CreateProjectMetadata">
```
Comment (davidfowl) [inline src/Aspire.Hosting/build/Aspire.Hosting.targets:66]: @baronfel can you give me an after the fact review on this?

Comment (baronfel) [reply src/Aspire.Hosting/build/Aspire.Hosting.targets:66]: This all seems ok to me, seems straightforward enough based on the existing patterns.

**Pull request merged**